File: examples/ch13/concept_template_method.py
"""Template Method — canonical conceptual example (runnable).

Fix an algorithm's skeleton in a base method; let subclasses fill in the steps
(and optionally override hooks). The skeleton's order is the invariant subclasses
cannot change.
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class AbstractFlow(ABC):
    def run(self) -> list[str]:
        """The template method: the fixed skeleton. Subclasses never override this."""
        return [self.step_one(), self.step_two(), self.hook()]

    @abstractmethod
    def step_one(self) -> str: ...

    @abstractmethod
    def step_two(self) -> str: ...

    def hook(self) -> str:
        """A hook: a default step, optional to override."""
        return "default"


class ConcreteFlow(AbstractFlow):
    def step_one(self) -> str:
        return "one"

    def step_two(self) -> str:
        return "two"


class CustomFlow(AbstractFlow):
    def step_one(self) -> str:
        return "1"

    def step_two(self) -> str:
        return "2"

    def hook(self) -> str:        # overrides the optional hook
        return "custom"


if __name__ == "__main__":
    print(ConcreteFlow().run())   # ['one', 'two', 'default']
    print(CustomFlow().run())     # ['1', '2', 'custom']

File: examples/ch13/ts/conceptTemplateMethod.ts
// Template Method — canonical conceptual example. Run: npx tsx conceptTemplateMethod.ts
//
// Fix an algorithm's skeleton in a base method; let subclasses fill in the steps
// (and optionally override hooks). The skeleton's order is the invariant subclasses
// cannot change.

export abstract class AbstractFlow {
  run(): string[] {
    // The template method: the fixed skeleton. Subclasses never override this.
    return [this.stepOne(), this.stepTwo(), this.hook()];
  }

  protected abstract stepOne(): string;
  protected abstract stepTwo(): string;

  protected hook(): string {
    // A hook: a default step, optional to override.
    return "default";
  }
}

export class ConcreteFlow extends AbstractFlow {
  protected stepOne(): string {
    return "one";
  }
  protected stepTwo(): string {
    return "two";
  }
}

export class CustomFlow extends AbstractFlow {
  protected stepOne(): string {
    return "1";
  }
  protected stepTwo(): string {
    return "2";
  }
  protected override hook(): string {
    return "custom";
  }
}

// Demo
console.log(new ConcreteFlow().run()); // ["one", "two", "default"]
console.log(new CustomFlow().run()); // ["1", "2", "custom"]
